Change dREG container

eternal-flame-AD opened the pull request:
Now dREG works. All minor CUDA versions (since CUDA Container was supported) are built and pushed

Container source: https://github.com/Functional-Genomics-Lab/containers/tree/main/src/dreg

I don't have access to the funlabs/ namespace on Docker Hub, may want to upload it there

Some necessary patches are applied on the upstream code, I explained them in the README above

We may not need the full samtools+bedops suite, but it's <20% of the base image size so I'm not too worried about it, just a nice to have.

I have tested the 12.5 container with this updated nextflow process and was able to run it.

<details><summary>Tree of /results/dreg</summary>

```
yume@tsubasa ~/s/dREG-nf (working-dreg) > tree results/dreg/                                                                                          (base) 06:59:37
results/dreg/
├── dREG_output_cd4_REP1_.chrom.info
├── dREG_output_cd4_REP1_.dREG.infp.bed.gz
├── dREG_output_cd4_REP1_.dREG.infp.bed.gz.tbi
├── dREG_output_cd4_REP1_.dREG.infp.bw
├── dREG_output_cd4_REP1_.dREG.peak.full.bed.gz
├── dREG_output_cd4_REP1_.dREG.peak.full.bed.gz.tbi
├── dREG_output_cd4_REP1_.dREG.peak.prob.bed.gz
├── dREG_output_cd4_REP1_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_cd4_REP1_.dREG.peak.prob.bw
├── dREG_output_cd4_REP1_.dREG.peak.score.bed.gz
├── dREG_output_cd4_REP1_.dREG.peak.score.bed.gz.tbi
├── dREG_output_cd4_REP1_.dREG.peak.score.bw
├── dREG_output_cd4_REP1_.dREG.raw.peak.bed.gz
├── dREG_output_cd4_REP1_.dREG.raw.peak.bed.gz.tbi
├── dREG_output_cd4_REP2_.chrom.info
├── dREG_output_cd4_REP2_.dREG.infp.bed.gz
├── dREG_output_cd4_REP2_.dREG.infp.bed.gz.tbi
├── dREG_output_cd4_REP2_.dREG.infp.bw
├── dREG_output_cd4_REP2_.dREG.peak.full.bed.gz
├── dREG_output_cd4_REP2_.dREG.peak.full.bed.gz.tbi
├── dREG_output_cd4_REP2_.dREG.peak.prob.bed.gz
├── dREG_output_cd4_REP2_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_cd4_REP2_.dREG.peak.prob.bw
├── dREG_output_cd4_REP2_.dREG.peak.score.bed.gz
├── dREG_output_cd4_REP2_.dREG.peak.score.bed.gz.tbi
├── dREG_output_cd4_REP2_.dREG.peak.score.bw
├── dREG_output_cd4_REP2_.dREG.raw.peak.bed.gz
├── dREG_output_cd4_REP2_.dREG.raw.peak.bed.gz.tbi
├── dREG_output_cd4_REP3_.chrom.info
├── dREG_output_cd4_REP3_.dREG.infp.bed.gz
├── dREG_output_cd4_REP3_.dREG.infp.bed.gz.tbi
├── dREG_output_cd4_REP3_.dREG.infp.bw
├── dREG_output_cd4_REP3_.dREG.peak.full.bed.gz
├── dREG_output_cd4_REP3_.dREG.peak.full.bed.gz.tbi
├── dREG_output_cd4_REP3_.dREG.peak.prob.bed.gz
├── dREG_output_cd4_REP3_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_cd4_REP3_.dREG.peak.prob.bw
├── dREG_output_cd4_REP3_.dREG.peak.score.bed.gz
├── dREG_output_cd4_REP3_.dREG.peak.score.bed.gz.tbi
├── dREG_output_cd4_REP3_.dREG.peak.score.bw
├── dREG_output_cd4_REP3_.dREG.raw.peak.bed.gz
├── dREG_output_cd4_REP3_.dREG.raw.peak.bed.gz.tbi
├── dREG_output_cd4_REP4_.chrom.info
├── dREG_output_cd4_REP4_.dREG.infp.bed.gz
├── dREG_output_cd4_REP4_.dREG.infp.bed.gz.tbi
├── dREG_output_cd4_REP4_.dREG.infp.bw
├── dREG_output_cd4_REP4_.dREG.peak.full.bed.gz
├── dREG_output_cd4_REP4_.dREG.peak.full.bed.gz.tbi
├── dREG_output_cd4_REP4_.dREG.peak.prob.bed.gz
├── dREG_output_cd4_REP4_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_cd4_REP4_.dREG.peak.prob.bw
├── dREG_output_cd4_REP4_.dREG.peak.score.bed.gz
├── dREG_output_cd4_REP4_.dREG.peak.score.bed.gz.tbi
├── dREG_output_cd4_REP4_.dREG.peak.score.bw
├── dREG_output_cd4_REP4_.dREG.raw.peak.bed.gz
├── dREG_output_cd4_REP4_.dREG.raw.peak.bed.gz.tbi
├── dREG_output_jurkat_REP1_.chrom.info
├── dREG_output_jurkat_REP1_.dREG.infp.bed.gz
├── dREG_output_jurkat_REP1_.dREG.infp.bed.gz.tbi
├── dREG_output_jurkat_REP1_.dREG.infp.bw
├── dREG_output_jurkat_REP1_.dREG.peak.full.bed.gz
├── dREG_output_jurkat_REP1_.dREG.peak.full.bed.gz.tbi
├── dREG_output_jurkat_REP1_.dREG.peak.prob.bed.gz
├── dREG_output_jurkat_REP1_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_jurkat_REP1_.dREG.peak.prob.bw
├── dREG_output_jurkat_REP1_.dREG.peak.score.bed.gz
├── dREG_output_jurkat_REP1_.dREG.peak.score.bed.gz.tbi
├── dREG_output_jurkat_REP1_.dREG.peak.score.bw
├── dREG_output_jurkat_REP1_.dREG.raw.peak.bed.gz
├── dREG_output_jurkat_REP1_.dREG.raw.peak.bed.gz.tbi
├── dREG_output_jurkat_REP2_.chrom.info
├── dREG_output_jurkat_REP2_.dREG.infp.bed.gz
├── dREG_output_jurkat_REP2_.dREG.infp.bed.gz.tbi
├── dREG_output_jurkat_REP2_.dREG.infp.bw
├── dREG_output_jurkat_REP2_.dREG.peak.full.bed.gz
├── dREG_output_jurkat_REP2_.dREG.peak.full.bed.gz.tbi
├── dREG_output_jurkat_REP2_.dREG.peak.prob.bed.gz
├── dREG_output_jurkat_REP2_.dREG.peak.prob.bed.gz.tbi
├── dREG_output_jurkat_REP2_.dREG.peak.prob.bw
├── dREG_output_jurkat_REP2_.dREG.peak.score.bed.gz
├── dREG_output_jurkat_REP2_.dREG.peak.score.bed.gz.tbi
├── dREG_output_jurkat_REP2_.dREG.peak.score.bw
├── dREG_output_jurkat_REP2_.dREG.raw.peak.bed.gz
└── dREG_output_jurkat_REP2_.dREG.raw.peak.bed.gz.tbi

1 directory, 84 files
```

</details>
Please review my nextflow code! This is my first real attempt at fixing nextflow.

Signed-off-by: eternal-flame-AD <yume@yumechi.jp>

eternal-flame-AD committed:
Change dREG container

Signed-off-by: eternal-flame-AD <yume@yumechi.jp>

edmundmiller commented on modules/local/dreg/main.nf:
```diff
@@ -1,7 +1,7 @@
+    publishDir "${params.outdir}/dreg/", mode: 'copy', pattern: "dREG_output_*"
 
-    container "docker.io/biohpc/dreg"
+    container "${moduleDir}/choose_dreg_image.sh".execute().text.trim()
```
Wild, does this work? There's also https://www.nextflow.io/docs/latest/process.html#beforescript which runs on the host, but I'm guessing this would too...

eternal-flame-AD replied:
yeah there is a chicken and egg problem here.. You don't know which container to use until you know the CUDA version of the job runner but you can't spawn a job runner without knowing which container to use. Maybe I will allow the user to override with their own values. But this will work on local runners so I think we should keep this, maybe move it to the local profile in nextflow.config

eternal-flame-AD replied:
Maybe I think it would be reasonable to assume we would have nvcc on the launching node if we really want to be heuristic?

edmundmiller commented on nextflow.config:
```diff
@@ -14,10 +14,18 @@ profiles {
+
+        process {
+            withName: 'DREG_RUN' {
+                // TODO: actually detect how many GPUs are available
```
https://www.nextflow.io/docs/latest/process.html#accelerator

eternal-flame-AD replied:
It says it only works on AWS and GCP. Local/Slurm runners with multiple GPUs probably need to be manually scheduled （and it was starving my development machine of resources so I set it to sequential)... 

Even on AWS and GCP I think "accelerator" from what I understand from the docs is not a resource limit, it's just a suggestion on what # and kind of GPU to ask for when creating cloud instances: so if you have 20 jobs and accelerator 1 you get 20 parallel executions regardless of what you put in params
Nextflow doesn't support custom resource limit it seems.

eternal-flame-AD replied:
It's actually pretty annoying here, simply limiting maxFork won't do because you still need to manually pass gpu IDs to docker and I don't see an obvious way to ask nextflow to keep track of which job gets which GPU so you don't know which GPU to take when spawning the container. I was cooking something basically wrapping Java BlockingDeque. I can get the GPU IDs assigned to processes now but still working on how to reliably release it, there's not much docs on nextflow on how to modify the scheduling it seems

edmundmiller reviewed (commented):
Seems like pubssh is having issues so I can't login to the GPU machines to test it right this second 😞. I'll check back in an hour or so!

eternal-flame-AD committed:
Upload workflow changes

Signed-off-by: eternal-flame-AD <yume@yumechi.jp>

eternal-flame-AD commented:
@edmundmiller I mapped out the bigwig generation from Nascent-Flow, rewrote them in DSL2 and refactored them a bit but the commands are the same. I also introduced bowtie2 and bwa-mem support however I can't test them to complete without a cluster, I only know they run for now. Here's a graph for the progress: yellow="mostly copied but not tested" green="implemented and did work on"

![](https://raw.githubusercontent.com/Functional-Genomics-Lab/dREG-nf/bc642d5c74057555b7b8bdb540c177a942f7fe20/subworkflows/prep_bw/Nascent-Flow.svg)

eternal-flame-AD commented:
Can we assume we already have indices usually in this case since it's a subworkflow? If so a lot of the code could be simplified and saves a lot of headache

edmundmiller commented:
Awesome. Before I just did this I wanted to ask, can we split those new commits off and merge the original PR? I still haven't gotten a chance to try it myself, but you've already gotten results out of it, and I can make a follow up PR if I find anything whenever I do get around to it.

So what I'm proposing:
1. Reset this branch to https://github.com/Functional-Genomics-Lab/dREG-nf/pull/1/commits/a09aa8b62b1eda8b1dca1c6e1dcf3c318f70f1fe
2. Merge this PR
3. Start a new PR with https://github.com/Functional-Genomics-Lab/dREG-nf/pull/1/commits/bc642d5c74057555b7b8bdb540c177a942f7fe20

eternal-flame-AD commented:
@edmundmiller Done!
